Parallelise integration tests using ResourceT

Anviking opened the pull request:
# Issue Number

ADP-466 — both ADP-529 and ADP-528


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] Use `runResourceT` in each test scenario to ensure only the wallet created in it are deleted.
- [x] Re-write all `listWallets` expectations to check that the expected wallets are a _subset_ of the actual wallets, instead of checking for equality.
- [x] Enable `-j 8` in buildkite
 


# Comments

- Buildkite now takes ≈35 min instead of 50 min
- On my machine `stack test cardano-wallet:integration --ta '-j 8'` takes 380s instead of ~1800s.
- Haven't enabled parallel tests in hydra yet because of errors
- I now pushed `anviking/ADP-466/parallel` (parallel _everything_) directly here.


<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Cleanup wallets in tests using ResourceT

instead of removing all wallets in tearDown, to allow for future
parallelisation.

Anviking committed:
Make HW_WALLETS_05 work with parallel

Anviking committed:
Re-write "BYRON_LIST_03 - Shelley CLI does not list Byron wallet"

Anviking committed:
Re-write more Shelley CLI wallet tests

Anviking committed:
Also adjust "The same account and mnemonic wallet can live side-by-side" test

Anviking committed:
Add some docs on postWallet and cleanup

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/HWWallets.hs:
```diff
@@ -332,27 +336,29 @@ spec = describe "SHELLEY_HW_WALLETS" $ do
 
             -- both wallets are available
-            rl <- request @[ApiWallet] ctx
-                (Link.listWallets @'Shelley) Default Empty
```
Replaced this `listWallets` call with individually fetching the wallets once again.

IIRC, for some reason there were some wallets that hadn't already been cleaned up, that were messing up this test. And we need to adapt for when enabling `parallel` anyway.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/HWWallets.hs:
```diff
@@ -303,46 +305,39 @@ spec = describe "SHELLEY_CLI_HW_WALLETS" $ do
-                    (#name . #getApiT . #getWalletName)
-                    (`shouldBe` T.pack pubKeyWalName)
-                ]
+            wids `shouldContain` [mnemonicWal ^. walletId]
```
Now checking that the wallet ids of the listed wallets _contain_ these two, instead of asserting that the first and second wallet in the list have the right names.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/Wallets.hs:
```diff
@@ -307,12 +314,9 @@ spec = describe "SHELLEY_CLI_WALLETS" $ do
                 \ or 24 words are expected."
-            (Stdout o, Stderr e) <- listWalletsViaCLI @t ctx
-            o `shouldBe` "[]\n"
-            e `shouldBe` cmdOk
```
🤔 I guess this is one of the more sub-optimal changes.

We _could_ calculate the corresponding wid, and make sure that no wallet with that wid is created. That would not catch if a wallet with an unexpected/different wid were created (like the wid calculation changing without the test being updated).

Best coverage would come from still running this test in isolation, and keeping the `== []` check.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/Wallets.hs:
```diff
@@ -335,9 +339,6 @@ spec = describe "SHELLEY_CLI_WALLETS" $ do
                 \ words are expected."
-            (Stdout o, Stderr e) <- listWalletsViaCLI @t ctx
-            o `shouldBe` "[]\n"
-            e `shouldBe` cmdOk
```
Same as https://github.com/input-output-hk/cardano-wallet/pull/2191/files#r509169056

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/Wallets.hs:
```diff
@@ -456,18 +457,22 @@ spec = describe "SHELLEY_CLI_WALLETS" $ do
-        j <- expectValidJSON (Proxy @[ApiWallet]) out
+        let walFromThisTest w =
+                prefix `T.isPrefixOf` getWalletName (getApiT (view #name w))
+        j <- filter walFromThisTest <$> expectValidJSON (Proxy @[ApiWallet]) out
```
Same check as before, but we name the wallets in this test with a unique prefix, such that we can filter out the rest.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/Wallets.hs:
```diff
@@ -480,14 +485,17 @@ spec = describe "SHELLEY_CLI_WALLETS" $ do
-        j <- expectValidJSON (Proxy @[ApiWallet]) out
+        let walFromThisTest w =
+                prefix `T.isPrefixOf` getWalletName (getApiT (view #name w))
+        j <- filter walFromThisTest <$> expectValidJSON (Proxy @[ApiWallet]) out
```
Similar `filter walFromThisTest` approach

Anviking committed:
Fix simplePayload and WALLET_UPDATE tests

Anviking committed:
Add counterexample to TRANS_EXTERNAL_02 to investigate failure

To investigate failures like:

```
  src/Test/Integration/Scenario/API/Shelley/Transactions.hs:1037:18:
  2) API Specifications, SHELLEY_TRANSACTIONS, TRANS_EXTERNAL_02 - Multiple Outputs Transaction - Shelley witnesses
       Waited longer than 90s (more than 2 epochs) to resolve action: "wFaucet and wSrc balances are as expected".
       expected: Quantity {getQuantity = 999989842531}
        but got: Quantity {getQuantity = 999989842430}
```

Anviking committed:
Add counterexample to TRANS_EXTERNAL_02 to investigate failure

To investigate failures like:

```
  src/Test/Integration/Scenario/API/Shelley/Transactions.hs:1037:18:
  2) API Specifications, SHELLEY_TRANSACTIONS, TRANS_EXTERNAL_02 - Multiple Outputs Transaction - Shelley witnesses
       Waited longer than 90s (more than 2 epochs) to resolve action: "wFaucet and wSrc balances are as expected".
       expected: Quantity {getQuantity = 999989842531}
        but got: Quantity {getQuantity = 999989842430}
```

Anviking committed:
Make OWASP_INJECTION_CREATE_WALLET_01 use a dynamic mnemonic

such that it works in parallel. Removing the `walletId` check shouldn't
really matter.

Anviking committed:
Add missing eventually to HW_WALLETS_03

To fix failures like:

```
 src/Test/Integration/Scenario/CLI/Shelley/HWWallets.hs:185:13:
  1) CLI Specifications, SHELLEY_CLI_HW_WALLETS, HW_WALLETS_03 - Cannot do operations requiring private key, Cannot send tx
       "Please enter your passphrase: **************\nI can't process this payment because there's not enough UTxO available in the wallet. The total UTxO sums up to 0 Lovelace, but I need 1000000 Lovelace (excluding fee amount) in order to proceed  with the payment.\n" does not contain "I couldn't find a root private key for the given wallet: 53a635ddc8caaee509f4bb991372bdbeeb4ebfde. However, this operation requires that I do have such a key.
```

Anviking committed:
Fix listWallets race condition

Fix maybeGetWallet to actually work using tryAnyDeep + runHandler

Anviking committed:
Add listFiltered{Byron,}Wallets helpers

Such that tests can list wallets only specific to themselves.

Theoretically we could extend `runResourceT` to keep track of this
automatically, and not manually have to pass in the wallet ids to
be included.

Anviking committed:
Adapt listWallets tests to work in parallel

Anviking committed:
Launch node with -N4 in integration tests

Anviking committed:
Enable `parallel` in integration tests

rvl reviewed (commented):
Wow, nice. Great results. I would say though that there needs to be at least _some_ tests running successfully in parallel for this to be worth merging. I assume the new failures are due to specs using a shared resource. Do you have any idea what that is? Stake pools perhaps?

I would also quite like this to stack a ReaderT on the ResourceT. Then we can make the `ctx` arg implicit everywhere. I think that will enhance readability by reducing boilerplate, if you follow through with a little bit of machinery. 

Check the `liftIO`s around verify and shouldBe, etc. I don't think they are needed if using hspec-expectations-lifted, right?

Anviking committed:
Rewrite WALLETS_GET_KEY_01 to a single test

Previously a `describe` with 3 `it`, each using the same wallet.

Race-conditions with the ResourceT cleanup... and simply parallel...
made this not work.

Let's simply convert them to a single test. We can use `counterexample`
to still show /which/ golden failed.

Anviking committed:
Make selectCoins work with MonadIO

Anviking committed:
Replace `liftIO $ verify` with `verify`

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Byron/Wallets.hs:
```diff
@@ -90,49 +88,50 @@ spec = describe "BYRON_CLI_WALLETS" $ do
-            c4 `shouldBe` ExitFailure 1
-            T.unpack err4 `shouldContain` (errMsg409WalletExists wid)
-            out4 `shouldBe` mempty
+            liftIO $ do
```
Should de-indent

Anviking committed:
Fix potential race in CLI HW_WALLETS_04

Commit:
Regenerate nix

Anviking committed:
Fixups which are difficult to squash

Anviking committed:
Add HasCallStack to counterexample

Anviking committed:
Debug log for TRANS_EXTERNAL_02

Anviking committed:
Fix for flaky TRANS_EXTERNAL_02

By mistake I first added the fix to TRANS_EXTERNAL_01 instead of
TRANS_EXTERNAL_02. But I'm not keeping it on both to be safe.

Anviking committed:
Fix flaky TRANS_DELETE_01

Co-Authored-By: KtorZ <matthias.benkort@gmail.com>

Anviking committed:
Disable parallel tests in hydra for now

Thinking the following errors are just from an overloaded hydra...

 src/Test/Integration/Scenario/CLI/Miscellaneous.hs:102:34:
  1) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help on bad arg or param, mnemonic --size 15
       uncaught exception: IOException of type InvalidArgument
       fd:23: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help on bad arg or param/mnemonic --size 15/"

  src/Test/Integration/Scenario/CLI/Miscellaneous.hs:102:34:
  2) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help on bad arg or param, wallet list --port
       uncaught exception: IOException of type InvalidArgument
       fd:27: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help on bad arg or param/wallet list --port/"

  src/Test/Integration/Scenario/CLI/Miscellaneous.hs:102:34:
  3) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help on bad arg or param, wallet create
       uncaught exception: IOException of type InvalidArgument
       fd:28: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help on bad arg or param/wallet create/"

  src/Test/Integration/Scenario/CLI/Miscellaneous.hs:102:34:
  4) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help on bad arg or param, wallet
       uncaught exception: IOException of type InvalidArgument
       fd:23: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help on bad arg or param/wallet/"

  src/Test/Integration/Scenario/CLI/Miscellaneous.hs:102:34:
  5) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help on bad arg or param, create
       uncaught exception: IOException of type InvalidArgument
       fd:27: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help on bad arg or param/create/"

  src/Test/Integration/Scenario/CLI/Miscellaneous.hs:102:34:
  6) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help on bad arg or param, wallet crate name
       uncaught exception: IOException of type InvalidArgument
       fd:28: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help on bad arg or param/wallet crate name/"

  src/Test/Integration/Scenario/CLI/Miscellaneous.hs:102:34:
  7) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help on bad arg or param, network information --port
       uncaught exception: IOException of type InvalidArgument
       fd:24: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help on bad arg or param/network information --port/"

  src/Test/Integration/Scenario/CLI/Miscellaneous.hs:109:27:
  8) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help with, -h
       uncaught exception: IOException of type InvalidArgument
       fd:24: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help with/-h/"

  src/Test/Integration/Scenario/CLI/Miscellaneous.hs:109:27:
  9) No backend required, Miscellaneous CLI tests, COMMON_CLI_MISC, CLI_HELP - cardano-wallet shows help with, --help
       uncaught exception: IOException of type InvalidArgument
       fd:24: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/No backend required/Miscellaneous CLI tests/COMMON_CLI_MISC/CLI_HELP - cardano-wallet shows help with/--help/"

  src/Test/Integration/Scenario/CLI/Shelley/Addresses.hs:206:51:
  10) CLI Specifications, SHELLEY_CLI_ADDRESSES, ADDRESS_LIST_04 - False wallet ids, 41 chars hex
       uncaught exception: IOException of type InvalidArgument
       fd:514: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/CLI Specifications/SHELLEY_CLI_ADDRESSES/ADDRESS_LIST_04 - False wallet ids/41 chars hex/"

  src/Test/Integration/Scenario/CLI/Shelley/Wallets.hs:306:40:
  11) CLI Specifications, SHELLEY_CLI_WALLETS, WALLETS_CREATE_05 - Can't create wallet with wrong size of mnemonic, 9
       uncaught exception: IOException of type InvalidArgument
       fd:594: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/CLI Specifications/SHELLEY_CLI_WALLETS/WALLETS_CREATE_05 - Can't create wallet with wrong size of mnemonic/9/"

  src/Test/Integration/Scenario/CLI/Shelley/Wallets.hs:418:50:
  12) CLI Specifications, SHELLEY_CLI_WALLETS, WALLETS_CREATE_08 - --address-pool-gap values, arbitraty string -> fail
       uncaught exception: IOException of type InvalidArgument
       fd:633: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/CLI Specifications/SHELLEY_CLI_WALLETS/WALLETS_CREATE_08 - --address-pool-gap values/arbitraty string -> fail/"

  src/Test/Integration/Scenario/CLI/Shelley/HWWallets.hs:212:9:
  13) CLI Specifications, SHELLEY_CLI_HW_WALLETS, HW_WALLETS_04 - Can manage HW wallet the same way as others, Can update name
       uncaught exception: IOException of type InvalidArgument
       fd:675: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/CLI Specifications/SHELLEY_CLI_HW_WALLETS/HW_WALLETS_04 - Can manage HW wallet the same way as others/Can update name/"

  src/Test/Integration/Scenario/CLI/Shelley/HWWallets.hs:379:46:
  14) CLI Specifications, SHELLEY_CLI_HW_WALLETS, HW_WALLETS_06 - Test parameters, Address pool gap invalid, Pool gap: 9
       uncaught exception: IOException of type InvalidArgument
       fd:700: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/CLI Specifications/SHELLEY_CLI_HW_WALLETS/HW_WALLETS_06 - Test parameters/Address pool gap invalid/Pool gap: 9/"

  src/Test/Integration/Scenario/CLI/Port.hs:133:5:
  15) CLI Specifications, COMMON_CLI_PORTS, PORT_03 - Cannot omit --port when server uses random port (wallet create)
       uncaught exception: IOException of type InvalidArgument
       fd:643: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/CLI Specifications/COMMON_CLI_PORTS/PORT_03 - Cannot omit --port when server uses random port (wallet create)/"

  src/Test/Integration/Scenario/CLI/Port.hs:184:5:
  16) CLI Specifications, COMMON_CLI_PORTS, PORT_03 - Cannot omit --port when server uses random port (address list)
       uncaught exception: IOException of type InvalidArgument
       fd:694: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/CLI Specifications/COMMON_CLI_PORTS/PORT_03 - Cannot omit --port when server uses random port (address list)/"

  src/Test/Integration/Scenario/CLI/Port.hs:196:13:
  17) CLI Specifications, COMMON_CLI_PORTS, PORT_04 - Fail nicely when port is out-of-bounds, serve --port 1010
       uncaught exception: IOException of type InvalidArgument
       fd:642: hGetContents: invalid argument (invalid byte sequence)

  To rerun use: --match "/CLI Specifications/COMMON_CLI_PORTS/PORT_04 - Fail nicely when port is out-of-bounds/serve --port 1010/"

rvl commented on nix/haskell.nix:
```diff
@@ -92,6 +92,14 @@ let
+          # Force more integration tests to run in parallel than the
+          # default number of build cores.
+          #
+          # TODO: Figure out why this doesn't work in hydra when it works in buildkite,
+          # and enable.
```
The `-j3` flag _should_ be added by this option when running `integration`.
Perhaps there are other things going on. Nix has two settings controlling parallelism - num cores for a build and num parallel builds - perhaps these are overriding our setting.
We can sort it later.

rvl reviewed (approved):
Excellent work.

Anviking committed:
Add counterexamples to debug TRANS_TTL_{01,02}
